### PR TITLE
chore: Update CI config

### PR DESCRIPTION
**Pull request by marinelli**
This PR proposes to:
- Clean-up `cabal.haskell-ci` removing all the ones with a default value
- Regenerate `haskell-ci.yml` workflow with `cabal.project`, and using the latest version of `haskell-ci`
- Configure `formatting-linting.yml` to reduce execution time and bump `fourmolu-action` release
- Bump ghc-9.2 and ghc-9.4 releases, and introduce ghc-9.6 build
- Refactor `fix-formatting.sh` and filter auto-generated files when formatting the code
- Minor Haskell changes due to `fourmolu` latest release
- Minor changes to cabal files
- Remove dependencies not required for building the library

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -1 +1,7 @@
 packages: . example
+
+source-repository-package
```
Comment (Siprj) [inline cabal.project]: Not sure what this is supposed to achieve. :thinking:
The Hackage version is fine, I think.

Comment (marinelli) [reply cabal.project]: The only changes we have there are for relaxing the constraints for building with GHC 9.6, we might replace it with some `allow-newer` declarations.

Comment (marinelli) [reply cabal.project]: done

**Review by Siprj — APPROVED** [review]
Looks good

**Pull request merged**